core/os/midi_driver.cpp

```diff
@@ -42,77 +42,82 @@ void MIDIDriver::set_singleton() {
 	singleton = this;
 }
 
-void MIDIDriver::receive_input_packet(int device_index, uint64_t timestamp, uint8_t *data, uint32_t length) {
-	Ref<InputEventMIDI> event;
-	event.instantiate();
-	event->set_device(device_index);
-	uint32_t param_position = 1;
-
-	if (length >= 1) {
-		if (data[0] >= 0xF0) {
-			// channel does not apply to system common messages
-			event->set_channel(0);
-			event->set_message(MIDIMessage(data[0]));
-			last_received_message = data[0];
-		} else if ((data[0] & 0x80) == 0x00) {
-			// running status
-			event->set_channel(last_received_message & 0xF);
-			event->set_message(MIDIMessage(last_received_message >> 4));
-			param_position = 0;
-		} else {
-			event->set_channel(data[0] & 0xF);
-			event->set_message(MIDIMessage(data[0] >> 4));
-			param_position = 1;
-			last_received_message = data[0];
-		}
-	}
+void MIDIDriver::receive_input_packet(int p_device_index, uint64_t p_timestamp, uint8_t *p_data, uint32_t p_length) {
+	// p_data may contain multiple messages, eg multiple note on using MIDI 'running status'.
+	uint32_t index = 0; // index will be incremented as each status or data byte is read from p_data.
 
-	switch (event->get_message()) {
-		case MIDIMessage::AFTERTOUCH:
-			if (length >= 2 + param_position) {
-				event->set_pitch(data[param_position]);
-				event->set_pressure(data[param_position + 1]);
-			}
-			break;
+	// Each time through loop will consume a MIDI message worth of bytes from p_data and create one InputEventMIDI.
+	while (index < p_length) {
+		Ref<InputEventMIDI> event;
+		event.instantiate();
+		event->set_device(p_device_index);
 
-		case MIDIMessage::CONTROL_CHANGE:
-			if (length >= 2 + param_position) {
-				event->set_controller_number(data[param_position]);
-				event->set_controller_value(data[param_position + 1]);
+		if (p_length - index >= 1) {
```

Comment:
Isn't this implied by the while loop?

Reply:
Yes but... The overall pattern in the method is that every time p_data is looked at to see if the next MIDI message can be read from it, there is this check to see if there are enough bytes for the whole message that is being checked for. I do not think that removing some of these >=1 checks  because of the while condition would improve code clarity. (I really intended to make a few necessary changes as needed to fix the bug, not rewrite the method.) I am however a beginner with how things are done at Godot.